### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by mtaillefumier**
Anton,

This pull request contains a first try to index the wavefunctions in a fashion similar to the beta projectors.  it compiles and pass all tests (including nio with QE). I also added a few bug fixes in the input files of the tests for the hubbard correction and renamed Hubbard_potential to Hubbard.
This pull request also contains a few additional functions in the API to satisfy the cp2k integration. 

More importantly it should not break anything in your repository.

Best

M.

### CODE REVIEW

File: src/Hubbard/hubbard.hpp [hunk 1]
```diff
@@ -137,7 +121,8 @@ class Hubbard_potential // TODO: rename to Hubbard
     /// apply the S operator in the us pp case. Otherwise it makes a simple copy
     void Apply_S_operator(K_point&                    kp,
```
Comment (toxa81) [inline src/Hubbard/hubbard.hpp]: Apply_S_operator -> apply_S_operator

File: src/Hubbard/hubbard.hpp [hunk 2]
```diff
@@ -278,8 +263,13 @@ class Hubbard_potential // TODO: rename to Hubbard
 
         this->lmax_ = -1;
```
Comment (toxa81) [inline src/Hubbard/hubbard.hpp]: put to constructor or static initialization

File: src/Hubbard/hubbard_occupancies_derivatives.hpp [hunk 3]
```diff
@@ -57,10 +57,11 @@ void Hubbard_potential::compute_occupancies_derivatives(K_point& kp,
     dmatrix<double_complex> Phi_S_Psi(HowManyBands, this->number_of_hubbard_orbitals() * ctx_.num_spins());
     matrix<double_complex> dm(this->number_of_hubbard_orbitals() * ctx_.num_spins(),
                               this->number_of_hubbard_orbitals() * ctx_.num_spins());
-    Phi_S_Psi.zero();
 
     #ifdef __GPU
     if (ctx_.processing_unit() == GPU) {
+        Phi_S_Psi.allocate(memory_t::device);
```
Comment (toxa81) [inline src/Hubbard/hubbard_occupancies_derivatives.hpp]: rename to phi_s_psi, dphi_s_psi

File: src/Hubbard/hubbard_occupancies_derivatives.hpp [hunk 4]
```diff
@@ -5,9 +5,9 @@
 
 // gradient of beta projectors. Needed for the computations of the forces
 
-void Hubbard_potential::compute_occupancies_derivatives(K_point& kp,
-                                                        Q_operator<double_complex>& q_op, // overlap operator
-                                                        mdarray<double_complex, 6>& dn_)                // Atom we shift
+void Hubbard::compute_occupancies_derivatives(K_point& kp,
+                                              Q_operator<double_complex>& q_op, // overlap operator
+                                              mdarray<double_complex, 6>& dn_)                // Atom we shift
```
Comment (toxa81) [inline src/Hubbard/hubbard_occupancies_derivatives.hpp]: function arguments have two trailing underscores (dn__ in this example), class members one underscore

File: src/Hubbard/hubbard_occupancy.hpp [hunk 5]
```diff
@@ -536,3 +535,72 @@ inline void Hubbard_potential::symmetrize_occupancy_matrix()
...
+ * @param what__ string to set to "set" for initializing sirius
+ * occupancy tensor and "get" for retrieving it
+ * @param pointer to external occupancy tensor
+ * @param leading dimension of the outside tensor
+ * @return
+ * return the occupancy matrix if the first parameter is set to "get"
+ */
+
+void Hubbard::access_hubbard_occupancies(char  const* what__,
+                                                   double_complex*      occ__,
```
Comment (toxa81) [inline src/Hubbard/hubbard_occupancy.hpp]: argument alignment

File: src/Hubbard/hubbard_potential_energy.hpp [hunk 6]
```diff
@@ -317,3 +319,73 @@ void Hubbard_potential::calculate_hubbard_potential_and_energy_non_colinear_case
...
+ * @param what__ string to set to "set" for initializing sirius
+ * occupancy tensor and "get" for retrieving it
+ * @param pointer to external potential tensor
+ * @param leading dimension of the outside tensor
+ * @return
+ * return the potential if the first parameter is set to "get"
+ */
+
+void Hubbard::access_hubbard_potential(char  const* what__,
+                                                 double_complex*      occ__,
```
Comment (toxa81) [inline src/Hubbard/hubbard_potential_energy.hpp]: argument alignment

File: src/Unit_cell/hubbard_orbitals_info.hpp [hunk 7]
```diff
@@ -0,0 +1,271 @@
+#ifndef __HUBBARD_ORBITAL_INFO_HPP_
+#define __HUBBARD_ORBITAL_INFO_HPP_
+
+namespace sirius {
+// structure containing all informations about a specific hubbard orbital (including the index of the radial function)
+    class hubbard_orbital_info_ {
```
Comment (toxa81) [inline src/Unit_cell/hubbard_orbitals_info.hpp]: this is probably a "hubbard_orbital_descriptor"

**Comment by mtaillefumier** [comment]
typos fixed

File: src/Unit_cell/hubbard_orbitals_info.hpp (hunk 7 above)
Comment (mtaillefumier) [reply src/Unit_cell/hubbard_orbitals_info.hpp]: True. I changed the name of the class to hubbard_orbital_descriptor_.

**Comment by mtaillefumier** [comment]
I think it is complete now

**Pull request merged**